### ISSUE TITLE
Display task code on task page

### ISSUE THREAD

**Issue by eliorc** [allegroai/trains#233]
As I see it, this is a really low hanging fruit with much use.

Previously I have experience with using Sacred + Omniboard as experiment trackers, and one of the most used features for me was to view the actual code of the executed experimented. Especially from the standpoint of a team lead or other type of supervision when viewing experiments if you are not the code author, it is super useful to actually see it in the same context.

Example from Omniboard project

![](https://raw.githubusercontent.com/vivekratnavel/omniboard/master/docs/assets/screenshots/source-file-view.png)

**Comment by bmartinn** [comment]
Thanks @eliorc 
If this is GitHub/GitLab/Bitbucket what I'm thinking is just a link opening an iframe / tab with the exact entry point script / commit.
Bear in mind the UI does not have direct access to the hosted git repository, but it could generate links.

What do you think?

**Comment by eliorc** [comment]
I don't think that is a good idea, since trains supports diffs - I believe it's preferable to see the code along with the diffs as in the exact code that was executed

As this is rather simple (I would do it myself if I knew some front end code) why not just save the current script running and render it with some kind of nice highlighting in the FE?

**Comment by bmartinn** [comment]
Hi @eliorc 
1. If there is no git repo, the entire code (a.k.a single python script) is stored in "uncommitted changes". Although without syntax highlighting, this seems like what you are after, no ?
2. What do you mean by " it's preferable to see the code along with the diffs" ? Do you mean like the GitHub diff display? Or are you referring to experiment comparison ?

**Comment by eliorc** [comment]
This got too complicated :sweat_smile: 

All I mean is just show the code as it is. What I meant with diffs, is that if I have committed code, but I have some diffs and I execute, I want to see it with the diff applied so I will see the script exactly as it ran.

From my point of view as a team lead, I'm looking at the experiments my DSes run, and I want to kind of "code review" the experiment... So I just want to be able to see in the dashboard the code they executed. Exactly like you see in "uncommitted changes" when you have no git.

Bottom line, just to see the code as it is in a better suited section (not "Uncommitted Changes") let's say called "Source Code" and preferably with highlights.

**Comment by bmartinn** [comment]
Ohh, now I see what you mean.
That makes a lot of sense.
I'm linking to #81 as this is a UI feature, expanding on the current capabilities of the system.
I'll talk to the front-end guys on how exactly to achieve that assuming most of the code is stored on a third-party site (a.k.a GitHub BitBucket etc.)

**Comment by berkgercek** [comment]
Is there any plan to implement this? I agree with the original poster that this would be a great feature to have, and much more informative than just a diff, repo, and path. It saves opening 3-4 tabs just to hunt down the referenced script.

It also saves against the issue that untracked files are not included in a diff, and as such an untracked script used to upload data is effectively untraceable if the user never adds/pushes it.

**Comment by patrickdasilva-clearml** [comment]
Hi @eliorc , this issue had been left behind, I believe since that time, we've implemented this (and you probably already know about it). We're closing the issue. Thanks for contributing!
